Accessibility First Draft

jayhersk opened the pull request.

jayhersk committed:
Accessibility First Draft

cyrus- commented on src/program-editors.rst:
```diff
@@ -39,4 +39,41 @@ TODO: work on version control systems
+Background
+----------------------------
+
+Accessibility in computing refers to the development and design of both software and hardware systems so as to be usable by people with disabilities. This includes screen readers for those with visual impairments and alternative input devices for those with motor impairments. Additionally, advances in computing technology over the previous decades have enabled richer communication for individuals ...
```
* remove "so as to be usable" (unnecessary words)
* "have improved for assisting" -> "assist"

cyrus- replied:
consider including a table of examples with two columns, "Disability" and "Assistive Technologies", rather than listing examples in paragraph form

cyrus- commented on src/program-editors.rst:
```diff
@@ -39,4 +39,41 @@ TODO: work on version control systems
+
+Accessibility in computing refers to the development and design of both software and hardware systems so as to be usable by people with disabilities. This includes screen readers for those with visual impairments and alternative input devices for those with motor impairments. Additionally, advances in computing technology over the previous decades have enabled richer communication for individuals ...
+ 
+Although personal computers were first sold to the general public in the 1970s, the process of making this technology accessible began to become more common around 1990. For example, the IBM Screen Reader for DOS was created in 1986 (initially only for employees) :cite:`cooke`, and in 1995, Windows 95 became the first Windows OS to have built in accessibility features, rather than as add-ons :cit ...
```
Make this is a section called "History"

cyrus- commented on src/program-editors.rst:
```diff
@@ -39,4 +39,41 @@ TODO: work on version control systems
+ 
+Although personal computers were first sold to the general public in the 1970s, the process of making this technology accessible began to become more common around 1990. For example, the IBM Screen Reader for DOS was created in 1986 (initially only for employees) :cite:`cooke`, and in 1995, Windows 95 became the first Windows OS to have built in accessibility features, rather than as add-ons :cit ...
+ 
+Accessibility research also appeared around this time. Like many HCI topics, it first emerged from the field of human factors. In ‘Thirty-Something Million: Should They Be Exceptions?’, Vanderheiden highlights the need to consider individuals with impairments in product design and development :cite:`vanderheiden1990thirty`. Bergman and Johnson provide an overview of accessibility as it relates sp ...
```
Please format descriptions of specific papers in the annotated bibliography format used in `notation.rst`.

cyrus- commented on src/program-editors.rst:
```diff
@@ -39,4 +39,41 @@ TODO: work on version control systems
+ 
+Accessibility research also appeared around this time. Like many HCI topics, it first emerged from the field of human factors. In ‘Thirty-Something Million: Should They Be Exceptions?’, Vanderheiden highlights the need to consider individuals with impairments in product design and development :cite:`vanderheiden1990thirty`. Bergman and Johnson provide an overview of accessibility as it relates sp ...
+ 
+Modern work has continued to explore design philosophies of accessible computing. Notably, in ‘Ability-Based Design: Concept, Principles and Examples’, Wobbrock et al. propose a shift in the frame of reference for thinking about accessible interfaces from disability to ability, and encourages designers to think about and adapt to what users can do :cite:`wobbrock2011ability`.  The authors present ...
```
Same here + rather than the long list at the end of this paragraph, format it as a bulleted list

cyrus- commented on src/program-editors.rst:
```diff
@@ -39,4 +39,41 @@ TODO: work on version control systems
+Accessibility research also appeared around this time. Like many HCI topics, it first emerged from the field of human factors. In ‘Thirty-Something Million: Should They Be Exceptions?’, Vanderheiden highlights the need to consider individuals with impairments in product design and development :cite:`vanderheiden1990thirty`. Bergman and Johnson provide an overview of accessibility as it relates sp ...
+ 
+Modern work has continued to explore design philosophies of accessible computing. Notably, in ‘Ability-Based Design: Concept, Principles and Examples’, Wobbrock et al. propose a shift in the frame of reference for thinking about accessible interfaces from disability to ability, and encourages designers to think about and adapt to what users can do :cite:`wobbrock2011ability`.  The authors present ...
+ 
```
There needs to be an introduction to the field of accessible programming environments specifically -- the discussion above is great, but it covers accessibility generally.

jayhersk replied:
I'm not entirely sure how to address this-- I tried to describe the entire section better in an overview. I had a hard time finding resources that gave an overview of accessibility as it specifically relates to programming environments. Please let me know if there is a source I can include?

cyrus- replied:
I think it's fine for this section to be a history of accessibility in general, but the overview at the very top of this section should say a little more about accessibility for programming, e.g. that many of the techniques that are used for accessibility in computing are also relevant when the user is a programmer, but there are also some unique challenges. You can then outline the remainder of the article by saying you'll talk first about accessibility in general then on work that focuses on programming interfaces.

jayhersk replied:
Restructured Overview and History sections to address this, also renamed 'Background' to 'Accessible Design Philosophies'

cyrus- commented on src/program-editors.rst:
```diff
@@ -39,4 +39,41 @@ TODO: work on version control systems
+ 
+Visual Impairments
+----------------------------
+Many programming environments are not purely text based, and use layout, icons, color, and other visual indicators to communicate code structure and debug information; similarly programming courses often use diagrams to convey abstract concepts. Traditional screen readers may not convey all of this information to visually impaired developers :cite:`doustdar_2016`. Thus, prior research has analyze ...
```
Refer to "secondary notation" here (see the Notation chapter for a description, you might want to include an inter-compendium link when you use that phrase).

jayhersk replied:
I have added a sentence mentioning this, but I am having trouble adding an inter-compendium link given the documentation. I will do my best to include this in the final draft.

cyrus- commented on src/program-editors.rst:
```diff
@@ -39,4 +39,41 @@ TODO: work on version control systems
+----------------------------
+Many programming environments are not purely text based, and use layout, icons, color, and other visual indicators to communicate code structure and debug information; similarly programming courses often use diagrams to convey abstract concepts. Traditional screen readers may not convey all of this information to visually impaired developers :cite:`doustdar_2016`. Thus, prior research has analyze ...
+ 
+Behaviors of Blind Programmers
```
This should be a subsection of Visual Impairments (you can use a different character for the line below the section to make it a subsection, e.g. `~~~~~~~~` instead of `-----------`)

cyrus- commented on src/program-editors.rst:
```diff
@@ -39,4 +39,41 @@ TODO: work on version control systems
+ 
+Armaly et al. conducted a study comparing reading behaviors between blind and sighted programmers :cite:`armaly2017comparison`. They asked 12 blind developers to read and summarize code, tracking their cursor and mouse movements, and compared the observed behaviors to a previous eye tracking study. They found that both groups prioritized reading method signatures, and that blind programmers retur ...
+ 
+Assistive Software for Blind Programmers
```
This should also be a subsection of Visual Impairments

cyrus- commented on src/program-editors.rst:
```diff
@@ -39,4 +39,41 @@ TODO: work on version control systems
+ 
+Motor Impairments
+----------------------------
+Motor impairments also require input devices other than the traditional keyboard and mouse, which can be painful or impossible to use for some :cite:`saphra_2019`. Alternatives can include speech based interaction. For example, Begel and Graham evaluate a speech-based programming system called SPEED (SPEech EDitor) :cite:`begel2006assessment`.
```
You can cite the Saphra blog post as a paper. Please describe it in more detail.

cyrus- replied:
Please also describe the Begel 2006 paper in more detail.

cyrus- commented on src/program-editors.rst:
```diff
@@ -39,4 +39,41 @@ TODO: work on version control systems
+ 
+Writing Accessible Software
+----------------------------
+Despite decades of accessibility research, a large proportion of software systems, websites, and mobile apps are not accessible to those that rely on screen readers or other accessibility technologies. Prior work has tried to educate developers on making use of accessibility APIs in a variety of ways, for example, in undergraduate courses :cite:`ludi2007introducing`, or through the creation of ne ...
```
Please describe the papers here in more detail.

cyrus- reviewed (changes requested):
The writing is good, nice job. Main problem is that you haven't formatted it in the "annotated bibliography" style. Please look at `notation.rst` and format your second draft in that style. Please also take a look at the comments and make the requested small edits + provide the additional details, especially regarding voice controlled programming. Looking forward to the talk!

jayhersk committed:
draft 2 updates

jayhersk commented:
Hi, I have made the requested changes for the second draft. Overall, I tried to style the citations as in notation.rst. Please let me know if the writing or citation style is not fitting! Thanks.

cyrus- commented on src/program-editors.bib:
```diff
@@ -0,0 +1,44 @@
+  title={An exploratory study of blind software developers},
+  author={Mealin, Sean and Murphy-Hill, Emerson},
+  year={2012},
+  organization={IEEE}
```
venue is missing

cyrus- commented on src/program-editors.bib:
```diff
@@ -0,0 +1,44 @@
+  publisher={IEEE}
+}
+ 
+@misc{doustdar_2016, title={Parham Doustdar's Blog}, url={https://www.parhamdoustdar.com/2016/03/27/autobiography-blind-programmer/}, journal={Parham Doustdar's Blog}, author={Doustdar, Parham}, year={2016}, month={Mar}}
```
The title of this should be the title of the blog post, not just "Parham Doustdar's Blog"

cyrus- replied:
put newlines between items in a bibtex entry

cyrus- commented on src/program-editors.bib:
```diff
@@ -0,0 +1,44 @@
+@article{vanderheiden1990thirty,
+  title={Thirty-something million: should they be exceptions?},
+  author={Vanderheiden, Gregg C},
+  journal={Human factors},
```
capitalize "Human Factors"

cyrus- commented on src/program-editors.bib:
```diff
@@ -0,0 +1,44 @@
+  year={1990},
+}
+ 
+@misc{ada, title={What is the Americans with Disabilities Act (ADA)?}, url={https://adata.org/learn-about-ada}, journal={ADA National Network}, year={2019}, month={Nov}}
```
put newlines between items in a bibtex entry

cyrus- commented on src/program-editors.bib:
```diff
@@ -0,0 +1,44 @@
+ 
+@misc{ada, title={What is the Americans with Disabilities Act (ADA)?}, url={https://adata.org/learn-about-ada}, journal={ADA National Network}, year={2019}, month={Nov}}
+ 
+@misc{history, title={History of Accesibility Technology}, url={https://cs.stanford.edu/people/eroberts/courses/soco/projects/2005-06/accessibility/firstwave.html}, journal={Accessible Technology in the 21st Century}, author={Hernandez, Jonathan and Mahajan, Shiv and Victor, Sarah}, year={2005}}
```
put newlines between items in a bibtex entry

cyrus- commented on src/program-editors.bib:
```diff
@@ -0,0 +1,44 @@
+ 
+@misc{history, title={History of Accesibility Technology}, url={https://cs.stanford.edu/people/eroberts/courses/soco/projects/2005-06/accessibility/firstwave.html}, journal={Accessible Technology in the 21st Century}, author={Hernandez, Jonathan and Mahajan, Shiv and Victor, Sarah}, year={2005}}
+ 
+@misc{microsoft, title={Microsoft accessibility}, url={https://www.microsoft.com/en-us/accessibility/}, journal={Microsoft accessibility}}
```
put newlines between items in a bibtex entry

cyrus- commented on src/program-editors.rst:
```diff
@@ -39,4 +39,252 @@ TODO: work on version control systems
+Introduction and Overview
+----------------------------
+
+Accessibility in computing refers to the development and design of both software and hardware systems to be usable by people with disabilities. Broadly, "Providing accessibility means removing barriers that prevent people with disabilities from participating in substantial life activities, including the use of services, products, and information." :cite:`bergman1995towards`. This includes providi ...
```
For long quotations, it would be helpful to use the block quotation format that sphinx supports

cyrus- commented on src/program-editors.rst:
```diff
@@ -39,4 +39,252 @@ TODO: work on version control systems
+  .. bibliography:: program-editors.bib
+    :filter: key == 'ada'
+
+  It’s also important to note that the Americans with Disabilities Act (ADA) was passed in 1990.
```
Remove "It's also important to note that"

cyrus- commented on src/program-editors.rst:
```diff
@@ -39,4 +39,252 @@ TODO: work on version control systems
+  .. bibliography:: program-editors.bib
+    :filter: key == 'doustdar_2016'
+
+  A blog post explaining the experiences of a blind programmer.
```
Can you say a little more about the observations that the author made

cyrus- commented on src/program-editors.rst:
```diff
@@ -39,4 +39,252 @@ TODO: work on version control systems
+  
+  They conducted qualitative interviews with eight developers with varying amounts of experience. Significant findings are summarized below:
+
+  1. All particpants relied heavily on screen readers. Two participants noted that braille displays were useful for matching parenthesis. A wide range of IDEs were used. Debuggers were rarely used.
```
"particpants" -> "participants"

cyrus- commented on src/program-editors.rst:
```diff
@@ -39,4 +39,252 @@ TODO: work on version control systems
+  They conducted qualitative interviews with eight developers with varying amounts of experience. Significant findings are summarized below:
+
+  1. All particpants relied heavily on screen readers. Two participants noted that braille displays were useful for matching parenthesis. A wide range of IDEs were used. Debuggers were rarely used.
+  2. Participants relied heavily on documentation and keywords to understnad code structure. Participants often also used temporary text files to save variable names or quickly edit single blocks.
```
"understnad" -> "understand"

cyrus- commented on src/program-editors.rst:
```diff
@@ -39,4 +39,252 @@ TODO: work on version control systems
+  .. bibliography:: program-editors2.bib
+    :filter: key == 'stefik2011design'
+
+  Stefik et al. present an educational infrastructure for blind and visually impaired middle and high school students that consists of Sodbeans, an auditory programming environment, Hop, a programming language, and a multi-sensory curriculum. They conducted an empirical study at a programming summer camp with 12 blind participants to evaluate their framework. 
```
What were the results of their evaluation?

cyrus- commented on src/program-editors.rst:
```diff
@@ -39,4 +39,252 @@ TODO: work on version control systems
+  .. bibliography:: program-editors3.bib
+    :filter: key == 'choo2019examining'
+
+  Another potential barier is that non-disabled users may not know how a disabled user would experience their software. This paper explores using virtual reality (VR) to simulate vision impairments for accessibility testing.
```
"barier" -> "barrier"

cyrus- commented on src/program-editors2.bib:
```diff
@@ -0,0 +1,56 @@
+@misc{cooke, title={A History of Accessibility at IBM}, url={https://www.afb.org/aw/5/2/14760}, journal={A History of Accessibility at IBM}, author={Cooke, Annemarie}}
```
* put newlines between items in a bibtex entry
* in titles, you have to put capitalized items in curly braces, e.g. `{IBM}`, or LaTeX will sometimes change the capitalization

cyrus- commented on src/program-editors2.bib:
```diff
@@ -0,0 +1,56 @@
+@misc{sanchez2005blind,
+  title={Blind learners programming through audio},
+  author={S{\'a}nchez, Jaime and Aguayo, Fernando},
+  booktitle={CHI'05 extended abstracts on Human factors in computing systems},
```
* Is there a full paper about this same work or was it only an extended abstract?
* Capitalization

jayhersk replied:
There is some follow up work by the authors but it is not related to programming tasks specifically ("3D sound interactive environments for blind children problem solving skills", "3D sound interactive environments for problem solving"), so I chose to include this poster instead.

cyrus- commented on src/program-editors2.bib:
```diff
@@ -0,0 +1,56 @@
+@misc{potluri2019ai,
+  title={AI-Assisted UI Design for Blind and Low-Vision Creators},
+  author={Potluri, Venkatesh and Grindeland, Tadashi and Froehlich, Jon E and Mankoff, Jennifer},
+  booktitle={the ASSETS'19 Workshop: AI Fairness for People with Disabilities},
```
remove "the"

cyrus- commented on src/program-editors2.bib:
```diff
@@ -0,0 +1,56 @@
+@misc{stefik2011design,
+  title={On the design of an educational infrastructure for the blind and visually impaired in computer science},
+  author={Stefik, Andreas M and Hundhausen, Christopher and Smith, Derrick},
+  booktitle={Proceedings of the 42nd ACM technical symposium on Computer science education},
```
Capitalization

cyrus- commented on src/program-editors2.bib:
```diff
@@ -0,0 +1,56 @@
+  organization={ACM}
+}
+ 
+@misc{saphra_2019, title={What Does a Coder Do If They Can't Type?}, url={http://nsaphra.github.io/post/hands/}, journal={Objective Funk}, author={Saphra, Naomi}, year={2019}, month={Aug}}
```
put newlines between items in a bibtex entry

cyrus- commented on src/program-editors2.bib:
```diff
@@ -0,0 +1,56 @@
+@misc{begel2006assessment,
+  title={An assessment of a speech-based programming environment},
+  author={Begel, Andrew and Graham, Susan L},
+  booktitle={Visual LAnguages and Human-Centric Computing (VL/HCC'06)},
```
"LAnguages" -> "Languages"

cyrus- commented on src/program-editors3.bib:
```diff
@@ -0,0 +1,29 @@
+  title={Toward an equal opportunity web: Applications, standards, and tools that increase accessibility},
+  author={Moreno, Lourdes and Martinez, Paloma and Ruiz, Belen and Iglesias, Ana},
+  year={2010},
+  publisher={IEEE}
```
missing venue here

cyrus- commented on src/program-editors3.bib:
```diff
@@ -0,0 +1,29 @@
+}
+ 
+@misc{gonzalez2005platform,
+  title={Platform-Independent Accessibility API: accessible document object model},
```
might need `{API}` here (also check for other issues like this)

cyrus- reviewed (changes requested):
This is great! Take a look at the comments for your final draft. Grade will be sent by email.

jayhersk committed:
fix references, fix intro

jayhersk committed:
Add additional paper details, fix typos

jayhersk commented:
Added final commit: Fixed introduction, reformatted sections relevant to introduction, added additional paper details when needed, fix references, fix typos.

cyrus- commented on src/program-editors.rst:
```diff
@@ -39,18 +39,20 @@ TODO: work on version control systems
-Accessibility in computing refers to the development and design of both software and hardware systems to be usable by people with disabilities. Broadly, "Providing accessibility means removing barriers that prevent people with disabilities from participating in substantial life activities, including the use of services, products, and information." :cite:`bergman1995towards`. This includes providi ...
-
-Additionally, advances in computing technology over the previous decades have enabled richer communication for individuals with impairments. For example, augmentative and alternative communication (AAC) tools such as video calling or speech generating devices assist individuals with impairments in producing or comprehending speech.
+Accessibility in computing refers to the development and design of both software and hardware systems to be usable by people with disabilities. Many of the techniques used to create accessible software are also relevant to program editors, however, these present some challenges. For example, programming environements may include real-time debug information, use color to encode meaning, and contai ...
```
"Many of the..." -- this is an awkwardly phrased sentence with the "however, " part, should be two sentences

cyrus- reviewed (approved):
great! final grades will be sent via email soon
